table/block_based_table_reader.cc

```diff
@@ -2157,76 +2184,247 @@ void BlockBasedTableIterator<TBlockIter>::InitDataBlock() {
+    // tombstone. Note that range_tombstone_ is not a raw range tombstone
+    // returned from RangeDelAggregator, but a cooked one. See
+    // InitRangeTombstone().
+    if (range_tombstone_start_ == nullptr) {
+      // The range tombstone extends to the beginning of the sstable.
+      if (range_tombstone_seq_ == 0) {
+        // The range doesn't apply to the keys in the sstable. Return the
+        // entry.
+        return;
+      }
+      // The range tombstone is a deletion tombstone. Loop to allow moving to
+      // the previous file.
+      ResetDataIter();
       return;
     }
+
+    auto ukey = user_key();
+    if (icomp_.user_comparator()->Compare(ukey, *range_tombstone_start_) < 0) {
+      // The key is past the tombstone. Grab the tombstone covering the
+      // current key. The new tombstone might cover the existing key, so loop
+      // so that we can have the proper check for whether the tombstone covers
+      // the key and is a deletion tombstone or not.
+      InitRangeTombstone(ukey);
+      continue;
+    }
+    // The key is contained within the current tombstone.
+    if (range_tombstone_seq_ == 0) {
+      // The tombstone doesn't apply to the sstable. Return the entry.
+      return;
+    }
+
+    // Find the previous non-deleted key.
+    index_iter_->Seek(tombstone_internal_start_key());
+    if (!index_iter_->Valid()) {
+      index_iter_->SeekToLast();
+      if (!index_iter_->Valid()) {
+        ResetDataIter();
+        block_iter_points_to_real_block_ = false;
+        return;
+      }
+    }
+    InitDataBlock();
+    block_iter_.SeekForPrev(tombstone_internal_start_key());
+    if (Valid()) {
+      InitRangeTombstone(user_key());
+    }
+  }
+}
+
+template <class TBlockIter>
+void BlockBasedTableIterator<TBlockIter>::InitRangeTombstone(const Slice& target) {
+  if (range_del_agg_ == nullptr || file_meta_ == nullptr) {
+    return;
   }
 
-  // We could have check lower bound here too, but we opt not to do it for
-  // code simplicity.
+  auto tombstone = range_del_agg_->GetTombstone(target, file_meta_->largest_seqno);
+  range_tombstone_start_ = tombstone.first.start;
+  range_tombstone_end_ = tombstone.first.limit;
+  range_tombstone_seq_ = tombstone.second;
+
+  // Clear the start key if it is less than the smallest key in the
+  // sstable. This allows us to avoid comparisons during Prev() in the common
+  // case.
+  if (range_tombstone_start_ != nullptr &&
+      icomp_.user_comparator()->Compare(
+          *range_tombstone_start_, file_meta_->smallest.user_key()) < 0) {
+    range_tombstone_start_ = nullptr;
+  }
+  // Clear the end key if it is larger than the largest key in the
+  // sstable. This allows us to avoid comparisons during Next() in the common
+  // case.
+  if (range_tombstone_end_ != nullptr &&
+      icomp_.user_comparator()->Compare(
+          *range_tombstone_end_, file_meta_->largest.user_key()) > 0) {
+    range_tombstone_end_ = nullptr;
```

Comment:
I noticed cases where we set `range_tombstone_end_ == nullptr` with `range_tombstone_seq_ == 0`, then later open a file in an upper level containing a range tombstone that covers part of this file. Since we don't reinitialize tombstones when the aggregator changes, we don't notice it so won't be able to seek within this file.

It's a rare edge case and potentially hard to fix. Just recording my observation here - won't block this PR on it.

Reply:
That's a good point. As I'm sure you're aware, it doesn't affect correctness due to the top-level `ShouldDelete` check. I don't see an obvious way to fix it, and I suspect it is too much of an edge case to worry about.